### PR TITLE
Implementation of Badges

### PR DESCRIPTION
**Pull request by nikhilrayaprolu**

### CODE REVIEW

**Comment by nikhilrayaprolu** [comment]
image

**Comment by nikhilrayaprolu** [comment]
image


@spMohanty  @skbly7 @piotrekpasciak Badges design currently

File: app/jobs/badges_daily_job.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    max_rank = ranks.last
+    participant_ids, participant_ranks  = [], []
+    during_badge = 0
+    if user_rating_service.temporary
+      during_badge = 3
+    end
+    teams_participant_ids.each_with_index do |team, i|
+      team.each_with_index do |participant_id, j|
+        participant_ranks = ranks[i].to_f/max_rank.to_f
+        if max_rank < 100
```
Comment (piotrekpasciak) [inline app/jobs/badges_daily_job.rb:16]: This condition seems to have same code as max_rank < 250. Please remove it.

File: app/jobs/badges_daily_job.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+          end
+        elsif max_rank > 100 && max_rank < 250
+          if participant_ranks < 0.1
+            Participant.find_by(id: participant_id).add_badge(3 + during_badge)
+          elsif participant_ranks < 0.2
+            Participant.find_by(id: participant_id).add_badge(2 + during_badge)
+          elsif participant_ranks < 0.4
+            Participant.find_by(id: participant_id).add_badge(1 + during_badge)
+          end
+        elsif max_rank > 250 && max_rank < 1000
```
Comment (piotrekpasciak) [inline app/jobs/badges_daily_job.rb:32]: Make it `max_rank > 250` and remove `max_rank > 1000`

File: app/services/discourse/fetch_badges_meta_service.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  class FetchBadgesMetaService < ::Discourse::BaseService
+    def initialize
+      @client = prepare_http_client
+    end
+
+    def call
+      return failure('Discourse API client couldn\'t be properly initialized.') if client.nil?
+
+      response      = client.post(badges_meta_path)
+      puts response
```
Comment (piotrekpasciak) [inline app/services/discourse/fetch_badges_meta_service.rb:11]: Remove this debugger from code.

File: app/services/discourse/fetch_badges_meta_service.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      response_hash         = map_response_body_to_hash(response.body)
+      success(response_hash)
+    rescue Discourse::Error => e
+      Logger.new(::Discourse::BaseService::LOGGER_URL).error("Unable to retrieve latest topics - #{e.message}")
```
Comment (piotrekpasciak) [inline app/services/discourse/fetch_badges_meta_service.rb:15]: This error message is wrong, you're retrieving badges meta. Please update it.

File: app/services/discourse/fetch_badges_service.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    def initialize(previous_id)
+      @client = prepare_http_client
+      @previous_id = previous_id
+    end
+
+    def call
+      return failure('Discourse API client couldn\'t be properly initialized.') if client.nil?
+
+      response      = client.post(latest_badges_path)
+      puts response
```
Comment (piotrekpasciak) [inline app/services/discourse/fetch_badges_service.rb:12]: Remove debugger.

File: app/services/discourse/fetch_badges_service.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      response_hash         = map_response_body_to_hash(response.body)
+      success(response_hash)
+    rescue Discourse::Error => e
+      Logger.new(::Discourse::BaseService::LOGGER_URL).error("Unable to retrieve latest topics - #{e.message}")
```
Comment (piotrekpasciak) [inline app/services/discourse/fetch_badges_service.rb:16]: Update error message.

File: app/views/participants/_show_badges.erb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+<span class="aicrowd-badge-info" style="<%= style ||= '' %>">
```
Comment (piotrekpasciak) [inline app/views/participants/_show_badges.erb:1]: Are you sure you need `<%= style ||= '' %>`. Isn't `<%= style %>` going to be enough?

File: config/initializers/merit.rb [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+#   id: (badge_id = badge_id+1),
+#   name: 'just-registered'
+# }, {
+#   id: (badge_id = badge_id+1),
+#   name: 'best-unicorn',
+#   custom_fields: { category: 'fantasy' }
+# }].each do |attrs|
+#   Merit::Badge.create! attrs
+# end
+Merit::Badge.create!(
```
Comment (piotrekpasciak) [inline config/initializers/merit.rb:34]: Are you sure this should be run in initializers? Isn't that going to create new records whenever you start the server? Please use seed file with rake task.

File: db/schema.rb [hunk 9]
```diff
@@ -10,7 +10,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2020_03_27_142425) do
+ActiveRecord::Schema.define(version: 2020_04_04_154409) do
```
Comment (piotrekpasciak) [inline db/schema.rb:13]: Please update schema to include only changes from migrations that are in this PR.

File: spec/jobs/badges_daily_job_spec.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (piotrekpasciak) [inline spec/jobs/badges_daily_job_spec.rb:1]: Remove empty file.

File: config/initializers/merit.rb (hunk 8 above)
Comment (nikhilrayaprolu) [reply config/initializers/merit.rb:34]: Yes,
These are not persisted in the database. These are just plain Rails objects that act like ActiveModel Records. For more information: https://github.com/norman/ambry

**Comment by nikhilrayaprolu** [comment]
@piotrekpasciak made the requested changes

File: app/jobs/badges_daily_job.rb [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    max_rank = ranks.last
+    participant_ids, participant_ranks  = [], []
+    during_badge = 0
+    if user_rating_service.temporary
+      during_badge = 3
+    end
+    teams_participant_ids.each_with_index do |team, i|
+      team.each_with_index do |participant_id, j|
+        participant_ranks = ranks[i].to_f/max_rank.to_f
+        if max_rank < 250
```
Comment (piotrekpasciak) [inline app/jobs/badges_daily_job.rb:16]: What if `max_rank` is  250? Please change this line to `max_rank <= 250`.

**Pull request merged**